### PR TITLE
GIVCAMP-314: Upgrade to Next 14.2

### PR DESCRIPTION
**Pull request by sherakama**
# READY FOR REVIEW

# Summary
- Upgrades everything to the latest and greatest
- Keeps the same build process of statically generating everything on build (Instead of on-demand). We could move to an on-demand invalidation strategy now, however.
- Allows link prefetching
- Removes extra home page page.tsx file

# Review By (Date)
- When you can

# Criticality
- Normal

# Review Tasks

1. Review the build https://nextupgrade-momentum.netlify.app/
2. Click around and ensure things are pre-rendered and loading fast (no skellys)
3. Try out the back and forward buttons on Safari to ensure it still works
4. Log in to storyblok https://app.storyblok.com/#/me/spaces/1005200/
5. Open up a test page and select the nextupgrade-momentum URL for the preview
6. Validate the preview updates appropriately as you edit
7. Save a change to a published page (or publish a test page new)
8. Wait for the build to complete
9. Validate change shows up on live site
10. Make another change and repeat the process to be sure
11. Navigate to a 404 page like https://nextupgrade-momentum.netlify.app/this-page-does-not-exist
12. ... Anything else?

### CODE REVIEW

**Comment by sherakama** [comment]
First build is looking ok: 

https://nextupgrade-momentum.netlify.app/

Now I need to write out the things to test that were broken in the previous version...

* Safari/Back-button issue
* Build cache not fetching fresh content on rebuilds
* Prefetch on links breaking nav
* Home page not loading or requiring SSR
* ... @yvonnetangsu what else do you recall on at the top of your mind?

**Comment by yvonnetangsu** [comment]
> First build is looking ok:
> 
> https://nextupgrade-momentum.netlify.app/
> 
> Now I need to write out the things to test that were broken in the previous version...
> 
> * Safari/Back-button issue
> * Build cache not fetching fresh content on rebuilds
> * Prefetch on links breaking nav
> * Home page not loading or requiring SSR
> * ... @yvonnetangsu what else do you recall on at the top of your mind?

Those are the things I remember as well 🤔 I'll test reduced motion mode as well just in case - not that we won't test that to begin with 😆

**Comment by sherakama** [comment]
Add, storyblok editor page live edit

**Comment by yvonnetangsu** [comment]
> Add, storyblok editor page live edit

Oh yes, that 😂  And 404 page

**Comment by yvonnetangsu** [comment]
> First build is looking ok:
> 
> https://nextupgrade-momentum.netlify.app/
> 
> Now I need to write out the things to test that were broken in the previous version...
> 
> * Safari/Back-button issue
> * Build cache not fetching fresh content on rebuilds
> * Prefetch on links breaking nav
> * Home page not loading or requiring SSR
> * ... @yvonnetangsu what else do you recall on at the top of your mind?

Adding one more - building locally.

File: app/(storyblok)/[[...slug]]/layout.tsx
Comment (sherakama) [inline app/(storyblok)/[[...slug]]/layout.tsx:1]: Moved this to an optional catch all segment as it handles all the pages

[Optional Catch-all Segments](https://nextjs.org/docs/app/building-your-application/routing/dynamic-routes#optional-catch-all-segments)

File: app/api/revalidate/route.ts
Comment (sherakama) [inline app/api/revalidate/route.ts:1]: This was just for testing. We can remove it if it worries you.

File: app/sitemap.ts
Comment (sherakama) [inline app/sitemap.ts:31]: Typescript thing...

File: components/Logo/Logo.tsx
Comment (sherakama) [inline components/Logo/Logo.tsx:33]: These are working now!

File: components/TypeForm/PopUp.tsx
Comment (sherakama) [inline components/TypeForm/PopUp.tsx:26]: Button props no longer supported the ref prop so I just moved it to a getElementById...

File: netlify.toml
Comment (sherakama) [inline netlify.toml:42]: TODO: Remove before merge.

File: netlify.toml
Comment (sherakama) [inline netlify.toml:56]: TODO: Restore before merge

File: next.config.js
Comment (sherakama) [inline next.config.js:6]: These are no longer experimental in 14

File: package.json
Comment (sherakama) [inline package.json:35]: We can probably remove this now that I've finished with testing.

File: package.json
Comment (sherakama) [inline package.json:61]: There was an incompatibility with @netlify/plugin-nextjs version 5.2 that this branch fixes. Let's wait until that PR gets merged and a new release gets punched before merging this one.

File: app/page.tsx
Comment (sherakama) [inline app/page.tsx:1]: Removed this as the home page is now rendered by the `(storyblok)/[[...slug]]` route.

**Comment by yvonnetangsu** [comment]
@sherakama Thank you for this upgrade 🚀 ! I'm testing now - and things all seem to be working, but strangely, I do see the loading rectangles sometimes, e.g., when I go to a story, and hit the back button. Is this expected (that it can still happen)? It seems a little strange because when I hit the back button, it should be going to a cached page that I already visited 🤔

**Comment by yvonnetangsu** [comment]
It now preloads the images that are not lazy loaded - and it honors the fetch priority if it's set also! 🎉 😄 

Screenshot 2024-05-07 at 11 02 04 AM

**Comment by yvonnetangsu** [comment]
https://github.com/vercel/next.js/discussions/54799

Interesting - it doesn't preload images when they're inside the picture tag now (it used to in the discussion above). I guess that makes sense since we don't want it to preload all the different sizes for each image 👀

File: components/TypeForm/PopUp.tsx
Comment (yvonnetangsu) [reply components/TypeForm/PopUp.tsx:26]: Huh I wonder why they removed that support

**Review by yvonnetangsu — APPROVED** [review]
I've tested everything - desktop, mobile, publishing etc, and pulled down locally with npm ci. Everything seems to be work as before with some preloading which is great! Other than I still see the loading skeletons (most often when I hit the back button - I'm blessed with slow internet at home so maybe that's why I notice it), everything seems to be working as expected 😄 !

**Comment by yvonnetangsu** [comment]
And thank you for working on this @sherakama ! Glad we're all caught up with the Next.js + Netlify plugin version 🎉 😄

File: package.json
Comment (yvonnetangsu) [inline package.json:59]: If we could update eslint-config-next to match the version of Next.js we're using that'd be awesome 😄

Comment (sherakama) [reply package.json:59]: Good catch. Thanks.

**Comment by sherakama** [comment]
> ! Other than I still see the loading skeletons (most often when I hit the back button)

I'll dig into this a bit.

**Comment by sherakama** [comment]
Screenshot 2024-05-08 at 09 48 01
Screenshot 2024-05-08 at 09 47 09
Screenshot 2024-05-08 at 09 46 35

**Comment by sherakama** [comment]
https://docs.netlify.com/platform/caching/#default-caching-behavior

> However, responses coming from [Netlify Functions](https://docs.netlify.com/functions/overview/), [Edge Functions](https://docs.netlify.com/edge-functions/overview/), and [proxies](https://docs.netlify.com/routing/redirects/rewrites-proxies/) are not cached by default. Because these responses are dynamic, they may change without a new deploy and we don’t want to risk serving stale content. If you would like to cache responses from functions, edge functions, or proxies, you can [add cache control headers](https://docs.netlify.com/platform/caching/#supported-cache-control-headers) to customize the caching.

**Comment by sherakama** [comment]
After setting the revalidation period to 1 year I was able to get a hit on the netlify edge. 

Screenshot 2024-05-10 at 15 29 30

**Comment by sherakama** [comment]
I still get the loader once in a while too. Usually starts after I navigate back to the home page. Perhaps coincidental but I'll look into that pattern.

**Comment by yvonnetangsu** [comment]
> I still get the loader once in a while too. Usually starts after I navigate back to the home page. Perhaps coincidental but I'll look into that pattern.

Thank you @sherakama ! Overall this is working much better than before already. If you're comfortable deploying this then I'm ok with that. Please feel free to look into finetuning later 😄

**Comment by sherakama** [comment]
Let me check in on https://github.com/SU-SWS/netlify-plugin-vault-variables/pull/28

**Comment by sherakama** [comment]
@yvonnetangsu ok, this one is ready for your final approval. Please take one last pass through.

**Comment by yvonnetangsu** [comment]
> @yvonnetangsu ok, this one is ready for your final approval. Please take one last pass through.

Thank you @sherakama ! I'll do that 👍🏼

**Comment by yvonnetangsu** [comment]
I published without clearing cache and it did update 👍🏼 😄 
Screenshot 2024-05-14 at 11 31 41 AM

**Review by yvonnetangsu — APPROVED** [review]
Did another check on desktop and mobile. Tried publishing, editing locally and using the preview. Tried building locally. Reduced motion, regular animation - all working great. Thank you @sherakama 🚀 ! I think this can go live - if you could please revert some of TODO stuff before merging then this is official 😄 🎉  We probably want to bump it up a major version for the release eh?

**Comment by sherakama** [comment]
> We probably want to bump it up a major version for the release eh?

Yeah, that sounds right to me.

**Comment by sherakama** [comment]
Thanks for the review and help @yvonnetangsu this has been merged and I will leave it up to you for when you want to punch a release.

**Comment by yvonnetangsu** [comment]
Let's push this to PROD today 🎉  This way we can use the new stuff for the "filtering" work 😄

**Comment by sherakama** [comment]
> Let's push this to PROD today 🎉 This way we can use the new stuff for the "filtering" work 😄

Ok, you are the boss.

File: package.json
Comment (yvonnetangsu) [inline package.json:35]: @sherakama We can remove this from package.json now right 👀 ? We can do that in the release PR

Comment (sherakama) [reply package.json:35]: I left it in as we did run into a conflict with this plugin and our vault plugin so I was cautious about locking the versions. When we remove it, Netlify will keep it up to date for us but we run the risk of it introducing a build error without us changing anything.

Comment (sherakama) [reply package.json:35]: I can go either way. Build errors won't cause the site to go down but it may delay content getting published and require us to scramble to figure out whats going on in the future. The trade, is that we will have to periodically update the plugin as well.

Comment (yvonnetangsu) [reply package.json:35]: Oh I didn't realize - sorry I removed it in the release PR. I can put it back in 😄

Comment (sherakama) [reply package.json:35]: It is all a trade and I can live with either scenario. You're also on the call sheet if this site starts to have a sad so think about which option you'd like to go with.

Comment (yvonnetangsu) [reply package.json:35]: That makes sense - I'll put it back in my horizontal changemaker card PR, then it can go into the release for that work since we'll probably be ok for now 🤞🏼 Thanks @sherakama !

**Pull request merged**
